File: client/web/src/features/recovery/api/recovery-api.ts
import { apiClient } from "@/lib/api-client";
import type {
  RecoveryRedeemResponse,
  RecoveryMaterialResponse,
  DeviceDto,
  Envelope,
  RecoveryDto,
} from "@/types/api-types";

export function redeemRecoveryCode(
  email: string,
  recoveryCode: string,
  verifierProof: { loginSessionId: string; clientProof: string },
): Promise<RecoveryRedeemResponse> {
  return apiClient.post<RecoveryRedeemResponse>("/recovery/codes/redeem", {
    email,
    recoveryCode,
    verifierProof,
  });
}

export function getRecoveryMaterial(
  sessionId: string,
  replacementDevice: DeviceDto,
): Promise<RecoveryMaterialResponse> {
  return apiClient.post<RecoveryMaterialResponse>(
    `/recovery/session/${sessionId}/material`,
    { replacementDevice },
  );
}

export function completeRecovery(
  sessionId: string,
  replacementDeviceId: string,
  deviceEnvelope: Envelope,
  rotatedRecovery: RecoveryDto,
): Promise<{ userId: string }> {
  return apiClient.post<{ userId: string }>(`/recovery/session/${sessionId}/complete`, {
    replacementDeviceId,
    deviceEnvelope,
    rotatedRecovery,
  });
}

export function regenerateRecoveryCodes(
  rotatedRecovery: RecoveryDto,
): Promise<{ rotatedAt: string }> {
  return apiClient.post<{ rotatedAt: string }>("/recovery/codes/regenerate", {
    rotatedRecovery,
  });
}

export function cancelRecoverySession(sessionId: string): Promise<void> {
  return apiClient.post<void>(`/recovery/session/${sessionId}/cancel`, {});
}

File: client/android/core/src/main/java/io/github/pstanar/pstotp/core/api/RecoveryApi.kt
package io.github.pstanar.pstotp.core.api

import io.github.pstanar.pstotp.core.model.api.DeviceDto
import io.github.pstanar.pstotp.core.model.api.Envelope
import io.github.pstanar.pstotp.core.model.api.RecoveryDto
import io.github.pstanar.pstotp.core.model.api.RecoveryMaterialResponse
import io.github.pstanar.pstotp.core.model.api.RecoveryRedeemResponse
import org.json.JSONObject

/** Recovery flow API endpoints (public — no Bearer token required). */
class RecoveryApi(private val client: ApiClient) {

    suspend fun redeemRecoveryCode(
        email: String,
        recoveryCode: String,
        verifierProof: JSONObject,
    ): RecoveryRedeemResponse {
        val body = JSONObject().apply {
            put("email", email)
            put("recoveryCode", recoveryCode)
            put("verifierProof", verifierProof)
        }
        return client.postPublic("/recovery/codes/redeem", body) { RecoveryRedeemResponse.fromJson(it) }
    }

    suspend fun getRecoveryMaterial(
        sessionId: String,
        replacementDevice: DeviceDto,
    ): RecoveryMaterialResponse {
        val body = JSONObject().put("replacementDevice", replacementDevice.toJson())
        return client.postPublic("/recovery/session/$sessionId/material", body) {
            RecoveryMaterialResponse.fromJson(it)
        }
    }

    suspend fun regenerateCodes(rotatedRecovery: RecoveryDto) {
        val body = JSONObject().put("rotatedRecovery", rotatedRecovery.toJson())
        client.post("/recovery/codes/regenerate", body) { }
    }

    suspend fun completeRecovery(
        sessionId: String,
        replacementDeviceId: String,
        deviceEnvelope: Envelope,
        rotatedRecovery: RecoveryDto,
    ): org.json.JSONObject {
        val body = JSONObject().apply {
            put("replacementDeviceId", replacementDeviceId)
            put("deviceEnvelope", deviceEnvelope.toJson())
            put("rotatedRecovery", rotatedRecovery.toJson())
        }
        return client.postPublic("/recovery/session/$sessionId/complete", body) { it }
    }
}
